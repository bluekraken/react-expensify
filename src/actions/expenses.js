// Imports
import { v4 as uuidv4 } from "uuid";
import db from "../firebase/firebase";

// Add expense
const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
});

export const saveAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            amount = 0,
            createdOn = 0,
            notes = ""
        } = expenseData;

        const expense = { description, amount, createdOn, notes };

        const guid = uuidv4();

        return db.collection("expenses")
            .doc(guid)
            .set({ ...expense })
            .then(() => {
                dispatch(addExpense({
                    guid,
                    ...expense
                }));
            });
    };
};


// Edit expense
export const editExpense = (guid, updates) => ({
    type: "EDIT_EXPENSE",
    guid,
    updates
});

// Remove expense
export const removeExpense = ({ guid } = {}) => ({
    type: "REMOVE_EXPENSE",
    guid
});