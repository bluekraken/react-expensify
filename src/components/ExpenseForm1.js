// Imports
import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

// Return the expense form
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            notes: props.expense ? props.expense.notes : "",
            amount: props.expense ?  (props.expense.amount / 100).toString() : "",
            createdOn: props.expense ? moment(props.expense.createdOn) : moment(),
            calendarFocused: false,
            error: ""
        };
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdOn) => {
        if (createdOn) {
            this.setState(() => ({ createdOn }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(() => ({ notes }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Please provide a description and an amount!"}));
        } else {
            this.setState(() => ({ error: ""}));
            this.props.handleSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdOn: this.state.createdOn.valueOf(),
                notes: this.state.notes
            });
        }
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdOn}
                        onDateChange={this.onDateChange}
                        focused={this.state.dateFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea
                        placeholder="Enter an optional note for your expense"
                        value={this.state.notes}
                        onChange={this.onNotesChange}
                    >
                    </textarea>
                    <button>{this.props.expense ? "Edit expense" : "Add expense"}</button>
                    {this.state.error && <p>{this.state.error}</p>}
                </form>
            </div>
        )
    }
}