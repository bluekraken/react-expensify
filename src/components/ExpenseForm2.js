// Imports
import React from "react";
import moment from "moment";
import DatePicker from "react-date-picker";

// Return the expense form
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            notes: props.expense ? props.expense.notes : "",
            amount: props.expense ?  (props.expense.amount / 100).toString() : "",
            createdOn: props.expense ? moment(props.expense.createdOn) : moment(),
            error: ""
        };
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (value) => {
        if (value) {
            this.setState(() => ({ createdOn: moment(value) }));
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNotesChange = (e) => {
        const notes = e.target.value;
        this.setState(() => ({ notes }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Please provide a description and an amount!"}));
        } else {
            this.setState(() => ({ error: ""}));
            this.props.handleSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdOn: this.state.createdOn.valueOf(),
                notes: this.state.notes
            });
        }
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <DatePicker
                        calendarAriaLabel="Toggle calendar"
                        clearIcon={null}
                        dayAriaLabel="Day"
                        format="dd/MM/y"
                        monthAriaLabel="Month"
                        nativeInputAriaLabel="Date"
                        onChange={this.onDateChange}
                        required={true}
                        value={this.state.createdOn.toDate()}
                        yearAriaLabel="Year"
                    />
                    <textarea
                        placeholder="Enter an optional note for your expense"
                        value={this.state.notes}
                        onChange={this.onNotesChange}
                    >
                    </textarea>
                    <button>{this.props.expense ? "Edit expense" : "Add expense"}</button>
                    {this.state.error && <p>{this.state.error}</p>}
                </form>
            </div>
        )
    }
}