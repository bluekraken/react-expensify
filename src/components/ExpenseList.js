// Imports
import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import getFilteredExpenses from "../selectors/getFilteredExpenses";

// Return the expense list
const ExpenseList = (props) => (
    <div>
        <h2>Expense list</h2>
        {props.expenses.map((expense) => <ExpenseListItem key={expense.guid} {...expense} />)}
    </div>
);

const mapStateToProps = (state) => ({ expenses: getFilteredExpenses(state.expenses, state.filters) });

export default connect(mapStateToProps)(ExpenseList);