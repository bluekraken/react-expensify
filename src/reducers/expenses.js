// Default array
const expensesDefaultState = [];

// Return expenses reducer
export default (state = expensesDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [...state, action.expense];
        case "EDIT_EXPENSE":
            return state.map((expense) => {
                if (expense.guid === action.guid) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case "REMOVE_EXPENSE":
            return state.filter(({ guid }) => guid !== action.guid);
        case "SETUP_EXPENSES":
            return action.expenses;
        default:
            return state;
    }
};