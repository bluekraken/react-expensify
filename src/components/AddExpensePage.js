// Imports
import React from "react";
import { connect } from "react-redux";

import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

// Return the add expense page
export class AddExpensePage extends React.Component {
    handleSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/dashboard");
    }

    render() {
        return (
            <div className="content-container">
                <div className="page-header">
                    <h1 className="page-header__title">Add Expense</h1>
                </div>
                <ExpenseForm
                    handleSubmit={this.handleSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);