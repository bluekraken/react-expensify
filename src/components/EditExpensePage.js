// Imports
import React from "react";
import { connect } from "react-redux";

import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

// Return the edit expense page
export class EditExpensePage extends React.Component {
    handleSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.guid, expense);
        this.props.history.push("/dashboard");
    }

    handleRemove = () => {
        this.props.startRemoveExpense(this.props.expense.guid);
        this.props.history.push("/dashboard");
    }

    render() {
        return (
            <div className="content-container">
                <div className="page-header">
                    <h1 className="page-header__title">Save expense</h1>
                </div>
                <ExpenseForm
                        expense={this.props.expense}
                        handleSubmit={this.handleSubmit}
                    />
                <div className="form__remove">
                    <button className="button button--remove" onClick={this.handleRemove}>
                        Remove expense
                    </button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.guid === props.match.params.guid)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (guid, expense) => dispatch(startEditExpense(guid, expense)),
    startRemoveExpense: (guid) => dispatch(startRemoveExpense(guid))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);