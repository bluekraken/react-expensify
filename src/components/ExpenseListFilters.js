// Imports
import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";

import { setTextFilter, setSortBy, setStartDate, setEndDate } from "../actions/filters";

// Return the expense list filters
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortByChange = (e) => {
        this.props.setSortBy(e.target.value);
    }

    render() {
        return (
            <div className="content-container input-group">
                <div className="input-group__item">
                    <input
                        className="text-input"
                        id="textFilter"
                        placeholder="Search expenses"
                        type="text" value={this.props.filters.text}
                        onChange={this.onTextChange}
                    />
                </div>
                <div className="input-group__item">
                    <DateRangePicker
                        startDate={this.props.filters.startDate}
                        startDateId="my-start-date-id"
                        endDate={this.props.filters.endDate}
                        endDateId="my-end-date-id"
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        showClearDates={true}
                        displayFormat="DD/MM/YYYY"
                    />
                </div>
                <div className="input-group__item">
                    <select
                        className="select"
                        id="sortBy"
                        value={this.props.filters.sortBy}
                        onChange={this.onSortByChange}
                    >
                        <option value="amount">Amount</option>
                        <option value="date">Date</option>
                    </select>
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state) => ({ filters: state.filters });

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setSortBy: (value) => dispatch(setSortBy(value)),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);