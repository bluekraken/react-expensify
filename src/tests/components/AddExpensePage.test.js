import React from "react";
import { shallow } from "enzyme";
import { AddExpensePage } from "../../components/AddExpensePage";
import expenses from "../fixtures/expenses";

let saveAddExpense, history, wrapper;

beforeEach(() => {
    saveAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage saveAddExpense={saveAddExpense} history={history} />);
});

test("should render AddExpensePage correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("should handle addExpense", () => {
    wrapper.find("ExpenseForm").prop("handleSubmit")(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(saveAddExpense).toHaveBeenLastCalledWith(expenses[2]);
});