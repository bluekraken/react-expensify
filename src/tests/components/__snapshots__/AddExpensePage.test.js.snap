// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="page-header"
  >
    <h1
      className="page-header__title"
    >
      Add Expense
    </h1>
  </div>
  <ExpenseForm
    handleSubmit={[Function]}
  />
</div>
`;
